algo_trees/ref_trees.py

```diff
@@ -0,0 +1,264 @@
+#
+#   Binary Trees
+#
+
+'''
+Exercice 0 : definition de la classe arbre
+'''
+
+class BinTree:
+    def __init__(self, key, left, right):
+        self.key = key
+        self.left = left
+        self.right = right
+
+
+'''
+Exercice 1 : tailler un arbre
+'''
+
+def size(tree):
+    if tree is None :
+        return 0
+    else:
+        return 1 + size(tree.left) + size(tree.right)
+
+
+'''
+Exercice 2 : prends de la hauteur !
+'''
+
+def height(tree):
+    if tree is None:
+        return -1
+    else:
+        return 1 + max(height(tree.left), height(tree.right))
+
+
+'''
+Exercice 3 : compter les feuilles de l'arbre
+'''
+
+def nb_leaves(tree):
+    count = 0
+
+    if tree is None:
+        return 0
+
+    if tree.left is None and tree.right is None:
+        count += 1
+    if tree.left:
+        count += nb_leaves(tree.left)
+    if tree.right:
+        count += nb_leaves(tree.right)
+
+    return count
+
+
+'''
+Exercice 4 : recherche dans un arbre
+'''
+
+def search(tree, x):
+    if tree is None:
+        return False
+    else:
+        if x == tree.key:
+            return True
+        else:
+            res = search(tree.left, x)
+            if res is not False:
+                return res
+            else:
+                return search(tree.right, x)
```

Comment:
Ça peut aussi se simplifier pas mal en vrai:

```suggestion
            return search(tree.left, x) or search(tree.right, x)
```

Reply:
Thanks, fixed in the last commit